File: leetcode/16_3Sum_Closest.py
import unittest
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        closest = nums[0] + nums[1] + nums[2]
        for i in range(n - 2):
            j, k = i + 1, n - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if s == target:
                    return s
                elif s > target:
                    k -= 1
                else:
                    j += 1
                if abs(target - s) < abs(target - closest):
                    closest = s
        return closest

class Test3SumClosest(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().threeSumClosest([-1,2,1,-4], 1), 2)    

    def test_2(self):
        self.assertEqual(Solution().threeSumClosest([0,0,0], 1), 0)    

File: leetcode/16_3Sum_Closest.rs
struct Solution { }

impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        nums.sort();
        let n = nums.len();
        let mut closest = nums[0] + nums[1] + nums[2];
        for i in 0..n - 2 {
            let (mut j, mut k) = (i + 1, n - 1);
            while j < k {
                let s = nums[i] + nums[j] + nums[k];
                if s == target {
                    return target;
                } else if s > target {
                    k -= 1;
                } else {
                    j += 1;
                }
                if (s - target).abs() < (closest - target).abs() {
                    closest = s;
                }
            }
        }
        closest
    }
}

#[cfg(test)]
mod three_sum_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::three_sum_closest(vec![-1,2,1,-4], 1), 2);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::three_sum_closest(vec![0,0,0], 1), 0);
    }
}